### PR TITLE
Post Fetching and LIke Button functionality

### PR DESCRIPTION
**Pull request by yashrajoria3**
Api for fetching post from backend, and like button functionality

### CODE REVIEW

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 1]
```diff
@@ -1,17 +1,205 @@
...
+  Typography,
+} from "@mui/material";
+import React, { useEffect, useState } from "react";
 import PostsFeed from "~/app/_components/postsFeed/PostsFeed";
+import { type Comment, type ArticlesList } from "./types/types";
+import { trpc } from "~/server/client";
+import { handleError } from "~/app/_utils/handleError";
+
 const MyFeed = () => {
+  const { user } = useClerk();
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: move all the logic to custom hook. Name it useMyFeedPage

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 2]
```diff
@@ -1,17 +1,205 @@
...
+  const [likedPosts, setLikedPosts] = useState<string[]>([]);
+
   const [tab, setTab] = useState(0);
+  const [skip, setSkip] = useState(0);
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: use custom component for tab created. use enums insead of indexes

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 3]
```diff
@@ -1,17 +1,205 @@
...
+
+  const postMutation = trpc.post;
+  const likeMutation = trpc.likes;
+  const commentMutation = trpc.comments;
+
+  const {
+    data: postData,
+    isLoading: queryLoading,
+    error,
+  } = postMutation.getPosts.useQuery(
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: try to fetch 15 posts initially, then on intervals of 200vh fetch 5 more posts

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 4]
```diff
@@ -1,17 +1,205 @@
...
+import { handleError } from "~/app/_utils/handleError";
+
 const MyFeed = () => {
+  const { user } = useClerk();
+  const [likedPosts, setLikedPosts] = useState<string[]>([]);
+
   const [tab, setTab] = useState(0);
+  const [skip, setSkip] = useState(0);
+  const [posts, setPosts] = useState<ArticlesList[]>([]);
+  const [isLoadingInitial, setIsLoadingInitial] = useState(true);
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: try using just the constant isLoading from useQuery exported from the custom hook

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 5]
```diff
@@ -1,17 +1,205 @@
...
+  const [isLoadingMore, setIsLoadingMore] = useState(false);
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: for isLoadingMore use the skip value and return from hook

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 6]
```diff
@@ -1,17 +1,205 @@
...
+  const postMutation = trpc.post;
+  const likeMutation = trpc.likes;
+  const commentMutation = trpc.comments;
+
+  const {
+    data: postData,
+    isLoading: queryLoading,
+    error,
+  } = postMutation.getPosts.useQuery(
+    { skip, limit: 3 },
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: use constants for skip and limit from common config

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 7]
```diff
@@ -1,17 +1,205 @@
...
+    {
+      enabled: skip >= 0 && hasMorePosts,
+    },
+  );
+
+  const { data: likedPostData } = likeMutation.getUserLikes.useQuery(
+    { userEmail: user?.primaryEmailAddress?.emailAddress ?? "" },
+    {
+      enabled: !!user?.primaryEmailAddress,
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: use  enabled: !!user?.primaryEmailAddress?.emailAddress,

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 8]
```diff
@@ -1,17 +1,205 @@
...
+  ): Promise<{ liked: boolean }> => {
+    if (user?.primaryEmailAddress?.emailAddress) {
+      try {
+        const response = await likePostMutation.mutateAsync({
+          postId,
+          userEmail: user?.primaryEmailAddress?.emailAddress,
+        });
+
+        setPosts((prevPosts) =>
+          prevPosts.map((post) =>
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: keep the logic of setpost outside the setstate and pass the value to it

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 9]
```diff
@@ -1,17 +1,205 @@
...
+          userEmail: user?.primaryEmailAddress?.emailAddress ?? "",
+          name:
+            user?.firstName === null
+              ? (user?.unsafeMetadata?.name as string)
+              : user?.firstName,
+          content,
+          profile: user.imageUrl,
+        });
+
+        setPosts((prevPosts) =>
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: take the logic out

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 10]
```diff
@@ -1,17 +1,205 @@
...
+              : post,
+          ),
+        );
+      } catch (error) {
+        handleError(error);
+      }
+  };
+
+  useEffect(() => {
+    if (postData) {
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: move the logic to custom hook. Use useMemo to store and update the state values. Try to get rid of the useEffect

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 11]
```diff
@@ -1,17 +1,205 @@
...
+      setIsLoadingMore(false);
+    }
+
+    if (error) {
+      setIsLoadingInitial(false);
+      setIsLoadingMore(false);
+    }
+  }, [postData, error, skip]);
+
+  useEffect(() => {
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: try to get rid of the useeffect

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 12]
```diff
@@ -1,17 +1,205 @@
...
+    const { scrollTop, scrollHeight, clientHeight } = document.documentElement;
+    const bottomReached = scrollHeight - scrollTop - clientHeight < 100;
+
+    if (bottomReached && hasMorePosts && !isLoadingMore) {
+      setSkip((prev) => prev + 3);
+      setIsLoadingMore(true);
+    }
+  };
+
+  useEffect(() => {
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: create a common hook for scroll listener which will accept a callback and pass the handleScroll to the hook

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 13]
```diff
@@ -51,29 +239,68 @@ const MyFeed = () => {
...
-            style={{
-              padding: "10px",
-            }}
-          >
-            Discover Tab
-          </div>
-        )}
-      </Grid>
-    </>
+      {postData?.length != 0 ? (
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: simplify the logic something like this,

 const renderUI = useMemo(()=>{
    if(tab === 0){
      
    }
    if(tab === 1){}
  },[])

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 14]
```diff
@@ -51,29 +239,68 @@ const MyFeed = () => {
...
+                width: "100%",
+                height: "100%",
+                display: "flex",
+                flexDirection: "column",
+                justifyContent: "center",
+                alignItems: "center",
+                gap: "16px",
+              }}
+            >
+              <CircularProgress />
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: use title prop in CircularProgress

File: src/app/(with-sidebar)/myfeed/page.tsx [hunk 15]
```diff
@@ -51,29 +239,68 @@ const MyFeed = () => {
...
+      ) : (
+        <Box
+          sx={{
+            display: "flex",
+            justifyContent: "center",
+            alignItems: "center",
+            height: "100%",
+          }}
+        >
+          <Typography> No Posts Found</Typography>
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/page.tsx]: create common components for loading and no data scenarios

File: src/app/(with-sidebar)/myfeed/types/types.ts [hunk 16]
```diff
@@ -27,31 +27,58 @@ export type MenuItemList = {
   icon: React.ReactNode;
 };
 
-type ArticlesList = {
+type ArticleMedia = {
+  thumbnail_picture: string[];
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myfeed/types/types.ts]: use camel casing

File: src/app/_components/CommentSection/CommentSection.tsx [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+import { toast } from "react-toastify";
+
+const CommentSection: React.FC<CommentSectionProps> = ({
+  comments,
+  addComment,
+}) => {
+  const [newComment, setNewComment] = useState("");
+
+  const commentButtonClick = async () => {
+    if (newComment?.trim().length > 1000) {
```
Comment (thekalamanchofficial) [inline src/app/_components/CommentSection/CommentSection.tsx]: dont keep any limits to comment length. Show indentation for comments

File: src/app/_components/commentCard/CommentCard.tsx [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+              NameFontSize={14}
+            />
+            <Typography
+              variant="caption"
+              sx={{
+                textAlign: "right",
+                marginRight: "10px",
+              }}
+            >
+              2 days ago
```
Comment (thekalamanchofficial) [inline src/app/_components/commentCard/CommentCard.tsx]: remove hardcode value, use created time

**Review by thekalamanchofficial — CHANGES_REQUESTED** [review]

File: src/app/(with-sidebar)/myfeed/page.tsx (hunk 14 above)
Comment (yashrajoria3) [reply src/app/(with-sidebar)/myfeed/page.tsx]: title prop does not show title

File: src/app/(with-sidebar)/myfeed/page.tsx (hunk 3 above)
Comment (yashrajoria3) [reply src/app/(with-sidebar)/myfeed/page.tsx]: same issue as before, 200vh condition met before the posts are shown on screen , making the next call before

File: package.json [hunk 19]
```diff
@@ -27,18 +27,22 @@
     "@trpc/client": "^11.0.0-rc.446",
     "@trpc/react-query": "^11.0.0-rc.446",
     "@trpc/server": "^11.0.0-rc.446",
+    "date-fns": "^4.1.0",
```
Comment (thekalamanchofficial) [inline package.json]: Why was this new package added. We already have dayjs in this project

File: package.json [hunk 20]
```diff
@@ -27,18 +27,22 @@
...
     "dayjs": "^1.11.13",
     "geist": "^1.3.0",
+    "install": "^0.13.0",
```
Comment (thekalamanchofficial) [inline package.json]: purpose of this library?

File: package.json [hunk 21]
```diff
@@ -27,18 +27,22 @@
...
     "mongodb": "^6.9.0",
     "mongoose": "^8.7.0",
     "next": "^14.2.4",
     "next-svg": "^1.0.7",
+    "npm": "^10.9.2",
```
Comment (thekalamanchofficial) [inline package.json]: Did we update npm ? Why?

File: package.json [hunk 22]
```diff
@@ -27,18 +27,22 @@
...
     "react": "^18.3.1",
     "react-datepicker": "^7.4.0",
     "react-dom": "^18.3.1",
     "react-hook-form": "^7.53.0",
     "react-otp-input": "^3.1.1",
     "react-quill": "^2.0.0",
+    "react-spring": "^9.7.4",
```
Comment (thekalamanchofficial) [inline package.json]: Why did we add this?

File: prisma/schema.prisma [hunk 23]
```diff
@@ -1,3 +1,104 @@
...
+// // learn more about it in the docs: https://pris.ly/d/prisma-schema
+
+// // Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
+// // Try Prisma Accelerate: https://pris.ly/cli/accelerate-init
+
+// generator client {
+//   provider = "prisma-client-js"
+// }
+
+// datasource db {
```
Comment (thekalamanchofficial) [inline prisma/schema.prisma]: Remove all these commented lines

File: src/app/(with-sidebar)/myProfile/mocks/myProfileMock.ts [hunk 24]
```diff
@@ -19,55 +19,93 @@ const AUTHOR_TO_FOLLOW_MOCK = Array.from({ length: 5 }, () => ({
...
+    media: {
+      thumbnail_picture: ["https://picsum.photos/seed/post1/800/600"],
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myProfile/mocks/myProfileMock.ts]: Haven't fixed this naming

File: src/app/(with-sidebar)/myProfile/page.tsx [hunk 25]
```diff
@@ -1,20 +1,127 @@
...
 import mockData from "./mocks/myProfileMock";
+import { type ArticlesList } from "~/app/(with-sidebar)/myfeed/types/types";
+import { useClerk } from "@clerk/nextjs";
+import { trpc } from "~/server/client";
+import { handleError } from "~/app/_utils/handleError";
 
 const MyProfile = () => {
   const [activeTab, setActiveTab] = React.useState(MyProfileTabsEnum.MY_POSTS);
-
+  const [posts, setPosts] = useState<ArticlesList[]>([]);
```
Comment (thekalamanchofficial) [inline src/app/(with-sidebar)/myProfile/page.tsx]: all this should be moved to hook. Let's do that as part of separate PR

File: src/app/_components/CommentSection/CommentCard.tsx [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+  isReplying?: boolean;
+  setIsReplying?: (replying: boolean) => void;
+  setReplyingTo?: (replyingTo: string) => void;
+  setParentState?: () => void;
+  setReplyingState?: (state: Record<string, boolean>) => void;
+  replyingState?: Record<string, boolean>;
+  handleReply?: (comment: string) => void;
+  isChildren: boolean;
+};
+const CommentCard: React.FC<commentCardProps> = ({
```
Comment (thekalamanchofficial) [inline src/app/_components/CommentSection/CommentCard.tsx]: Did we need these comment components?

File: src/server/routers/post.ts [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+          .positive("Bid amount must be positive"),
+        createdAt: yup.date().optional(),
+      }),
+    )
+    .optional(),
+  hasMorePosts: yup.boolean().optional(),
+});
+
+export const postRouter = router({
+  // getPosts: publicProcedure
```
Comment (thekalamanchofficial) [inline src/server/routers/post.ts]: remove these commented line

File: src/app/(with-sidebar)/myfeed/page.tsx (hunk 12 above)
Comment (thekalamanchofficial) [reply src/app/(with-sidebar)/myfeed/page.tsx]: this is still pending

File: src/app/(with-sidebar)/myfeed/types/types.ts (hunk 16 above)
Comment (thekalamanchofficial) [reply src/app/(with-sidebar)/myfeed/types/types.ts]: still pending

Comment (yashrajoria3) [reply src/app/(with-sidebar)/myfeed/types/types.ts]: done

File: package.json (hunk 19 above)
Comment (yashrajoria3) [reply package.json]: deleted

File: package.json (hunk 20 above)
Comment (yashrajoria3) [reply package.json]: removed

File: package.json (hunk 21 above)
Comment (yashrajoria3) [reply package.json]: deleted

File: package.json (hunk 22 above)
Comment (yashrajoria3) [reply package.json]: removed

File: prisma/schema.prisma (hunk 23 above)
Comment (yashrajoria3) [reply prisma/schema.prisma]: done

File: src/app/(with-sidebar)/myProfile/mocks/myProfileMock.ts (hunk 24 above)
Comment (yashrajoria3) [reply src/app/(with-sidebar)/myProfile/mocks/myProfileMock.ts]: done

File: src/server/routers/post.ts (hunk 27 above)
Comment (yashrajoria3) [reply src/server/routers/post.ts]: done

**Review by thekalamanchofficial — APPROVED** [review]

**Pull request merged**